### PR TITLE
support kimi 2.5/6 lora (logprob diff exist)

### PR DESCRIPTION
**Pull request by nanjiangwill**
Adds Kimi K2.5 / K2.6 LoRA support to the \`sglang-miles\` branch:

- **MoE LoRA: \`experts_shared_outer_loras\` layout.** \`gate_up_proj_moe\` has shared A (\`(slot, 1, 2r, hidden)\`) + per-expert B; \`down_proj_moe\` has the inverse. \`LoRAMemoryPool\` initializes A and B independently per target so a shared Tensor on one side doesn't clobber the per-expert dict on the other.
- **Virtual-experts kernel: \`lora_b\` as a list.** \`merged_experts_fused_moe_lora_add\` accepts a length-2 \`lora_b\` for the gated case (one \`2r\` shrink, two \`r\` expands); \`lora_moe_runners\` passes \`tuple(lora_b_stacked)\` for it.
- **Chunked \`load_lora_adapter_from_tensors\`.** Per-rank serialization: \`serialized_tensors: str\` -> \`serialized_named_tensors: List[Union[str, bytes]]\` so each TP rank deserializes its own producer's CUDA-IPC bucket (matches \`update_weights_from_tensor\`). \`Engine\` and \`BaseTpWorker\` plumbed accordingly; \`TokenizerControlMixin\` relaxes \`dp_size == 1\` to also allow \`enable_dp_attention\`.
- **DP-attention safety.** Token→lora mapping pads \`req_to_lora\` with a \`-1\` sentinel so foreign tokens past \`seg_indptr[-1]\` don't \`searchsorted\` past the end / read uninitialised memory.
- **Quant post-process skips LoRA wrappers.** \`restore_weights_before_load\` / \`process_weights_after_loading\` loops in \`ModelRunner\` skip \`BaseLayerWithLoRA\`; the inner base layer is yielded separately by \`named_modules\` and gets processed there.
- **MLA fused-A cache persistence.** \`DeepseekV2WeightLoaderMixin\` persists \`cached_a_proj\` across \`do_load_weights\` calls so chunked updates that split \`q_a_proj\` / \`kv_a_proj_with_mqa\` across chunks still produce the fused \`fused_qkv_a_proj_with_mqa\` once both halves arrive. (Reverted on \`sglang-miles\` in sgl-project/sglang#28561 — sender-side pairing in radixark/miles#1361 is the cleaner fix.)
- **Base-layer dim derivation.** \`BaseLoRABackend\` derives MoE intermediate buffer dims from the base FusedMoE rather than quant-specific tensors so it works for any scheme (FP, WNA16, Marlin-packed, …).

\`logprob diff exist\` in the title flags a residual \`train/train_rollout_logprob_abs_diff\` drift seen on K2.5 RL runs;

---

**Decomposition for \`main\`:** sgl-project/sglang#28562, sgl-project/sglang#28564, sgl-project/sglang#28566. The MoE post-process LoRA-wrapper skip is blocked on sgl-project/sglang#28565 (upstream prereq). The MLA fused-A cache change is being reverted on \`sglang-miles\` via sgl-project/sglang#28561.

### CODE REVIEW

File: python/sglang/srt/lora/triton_ops/virtual_experts.py
Comment (yushengsu-thu) [inline python/sglang/srt/lora/triton_ops/virtual_experts.py:550]: Is this kimi specific modification? why need to do this?

File: python/sglang/srt/lora/layers.py [hunk 1]
```diff
@@ -826,6 +861,12 @@ def __init__(
             and base_layer.quant_method.runner is not None
         ):
             runner_backend = base_layer.quant_method.runner.runner_backend
+        elif (
+            hasattr(base_layer, "scheme")
+            and hasattr(base_layer.scheme, "runner")
+            and base_layer.scheme.runner is not None
+        ):
+            runner_backend = base_layer.scheme.runner.runner_backend
```
Comment (yushengsu-thu) [inline python/sglang/srt/lora/layers.py]: Did you cherry pick this modification from the upstream? Do you remember this?

File: python/sglang/srt/lora/mem_pool.py [hunk 2]
```diff
@@ -157,6 +157,66 @@ def _get_num_experts(base_model: torch.nn.Module) -> int:
             or 1
         )
 
+    def _row_parallel_shard_tp(
```
Comment (yushengsu-thu) [inline python/sglang/srt/lora/mem_pool.py:287]: This is kimi specific, right?

File: python/sglang/srt/managers/tp_worker.py [hunk 3]
```diff
@@ -193,19 +193,23 @@ def unload_lora_adapter(self, recv_req: UnloadLoRAAdapterReqInput):
...
+        # ``update_weights_from_tensor`` above. One producer per one
+        # consumer means the CUDA-IPC ref counter on the producer's
+        # bucket drops cleanly each cycle.
+        monkey_patch_torch_reductions()
+        serialized = recv_req.serialized_named_tensors[self.tp_rank]
         if recv_req.load_format == "flattened_bucket":
-            flattened_data = MultiprocessingSerializer.deserialize(
-                recv_req.serialized_tensors
-            )
+            flattened_data = MultiprocessingSerializer.deserialize(serialized)
```
Comment (yushengsu-thu) [inline python/sglang/srt/managers/tp_worker.py:206]: it might have problems when you run these two tests:
test:
examples/lora/run-qwen3-4b-megatron-lora-result.sh
examples/lora/run-qwen2.5-0.5B-megatron-lora.sh

**Comment by yushengsu-thu** [comment]
these modification should be sent to sglang/main as well or it will becomse useless after we upgrade the sgl

File: python/sglang/srt/lora/mem_pool.py (hunk 2 above)
Comment (nanjiangwill) [reply python/sglang/srt/lora/mem_pool.py:287]: for lora dp_attention > 1  general case

File: python/sglang/srt/managers/tp_worker.py (hunk 3 above)
Comment (nanjiangwill) [reply python/sglang/srt/managers/tp_worker.py:206]: tested

**Review by yushengsu-thu — APPROVED** [review]

**Comment by yushengsu-thu** [comment]
the modification are almost lora related only [‎python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py‎](https://github.com/sgl-project/sglang/pull/25141/changes#diff-e7b662c457f2c2d0176ebdbf595bd5936bace6ba129712cad3bdc8c00af701d5) is related to kimi2.5 base.  

We can consider if revert the kimi2.5 base related parts. To unblock kimi full-param and lora training support so I merge them first.
